pkg/translate/translate.go

```diff
@@ -0,0 +1,413 @@
+				nnp := append(newPath, key.String())
+				errs = util.AppendErrs(errs, t.insertLeaf(root, nnp, fieldValue.MapIndex(key)))
+			}
+		case reflect.Slice:
+			dbgPrint("Slice")
+			for i := 0; i < fieldValue.Len(); i++ {
+				errs = util.AppendErrs(errs, t.protoToValues(fieldValue.Index(i).Elem().Interface(), root, path))
+			}
+		case reflect.Ptr:
+			if util.IsNilOrInvalidValue(fieldValue.Elem()) {
+				continue
+			}
+			newPath := append(path, fieldName)
+			if fieldValue.Elem().Kind() == reflect.Struct {
+				dbgPrint("Struct Ptr")
+				errs = util.AppendErrs(errs, t.protoToValues(fieldValue.Interface(), root, newPath))
+			} else {
+				fmt.Println("Leaf Ptr")
+				errs = util.AppendErrs(errs, t.insertLeaf(root, newPath, fieldValue))
+			}
+		default:
+			dbgPrint("field has kind %s", kind)
+			if structElems.Field(i).CanInterface() {
+				errs = util.AppendErrs(errs, t.insertLeaf(root, append(path, fieldName), fieldValue))
+			}
+		}
+	}
+	return errs
+}
+
+func (t *Translator) insertLeaf(root util.Tree, newPath util.Path, fieldValue reflect.Value) (errs util.Errors) {
+	// Must be a scalar leaf. See if we have a mapping.
+	valuesPath, m := getValuesPathMapping(t.APIMapping, newPath)
+	var v interface{}
+	if fieldValue.Kind() == reflect.Ptr {
+		v = fieldValue.Elem().Interface()
+	} else {
+		v = fieldValue.Interface()
+	}
+	switch {
+	case m == nil:
+		break
+	case m.translationFunc == nil:
+		// Use default translation which just maps to a different part of the tree.
+		errs = util.AppendErr(errs, defaultTranslationFunc(m, root, valuesPath, v))
+	default:
+		// Use a custom translation function.
+		errs = util.AppendErr(errs, m.translationFunc(m, root, valuesPath, v))
+	}
+	return errs
+}
+
+// getValuesPathMapping tries to map path against the passed in mappings with a longest prefix match. If a matching prefix
+// is found, it returns the translated YAML path and the corresponding translation.
+// e.g. for mapping "a.b"  -> "1.2", the input path "a.b.c.d" would yield "1.2.c.d".
+func getValuesPathMapping(mappings map[string]*Translation, path util.Path) (string, *Translation) {
+	p := path
+	var m *Translation
+	for ; len(p) > 0; p = p[0 : len(p)-1] {
+		m = mappings[p.String()]
+		if m != nil {
+			break
+		}
+	}
+	if m == nil {
+		return "", nil
+	}
+
+	if m.inPath == "" {
+		return "", m
+	}
+
+	d := len(path) - len(p)
+	out := m.inPath + "." + path[len(path)-d:].String()
+	dbgPrint("translating %s to %s", path, out)
+	return out, m
+}
+
+// setTree sets the YAML path in the given Tree to the given value, creating any required intermediate nodes.
+func setTree(root util.Tree, path util.Path, value interface{}) error {
+	dbgPrint("setTree %s:%v", path, value)
+	if len(path) == 0 {
+		return fmt.Errorf("path cannot be empty")
+	}
+	if len(path) == 1 {
+		root[path[0]] = value
+		return nil
+	}
+	if root[path[0]] == nil {
+		root[path[0]] = make(util.Tree)
+	}
+	setTree(root[path[0]].(util.Tree), path[1:], value)
```

Comment:
it should handle the error of setting subtrees

Reply:
Done.